pandas/core/_meta.py

```diff
@@ -0,0 +1,146 @@
+
+dispatch = defaultdict(dict)
+dispatch_method_type = Union[Callable[..., "NDFrame"], str]
+
+
+def key_of(method):
+    if isinstance(method, str):
+        # TODO: figure out if this is OK. May be necessary when we have
+        #   things like pd.merge and DataFrame.merge that hit the same finalize.
+        return method
+    elif method:
+        return method.__module__, method.__name__
+
+
+class PandasMetadata:
+    """
+    Dispatch metadata finalization for pandas metadata.
+
+    Users should instantiate a single `PandasMetadata` instance
+    for their piece of metadata and register finalizers for various
+    pandas methods using :meth:`PandsaMetadata.register`.
+
+    Parameters
+    ----------
+    name : str
+        The name of the attribute being finalized.
+
+    Examples
+    --------
+    >>> maxmeta = PandasMetadata("attr")
+
+    Register a finalizer for a given pandas method:
+
+    >>> @maxmeta.register(pd.concat)
+    ... def _(new, concatenator):
+    ...     new.attr = max(x.attr_meta for x in concatenator.objs)
+
+    >>> pd.DataFrame._metadata = ['attr']
+    >>> x = pd.DataFrame({"x"}); x.attr = 1
+    >>> y = pd.DataFrame({"y"}); y.attr = 2
+    >>> pd.concat([x, y]).attr
+    2
+    """
+
+    def __init__(self, name: str):
+        self.name = name
+
+    def register(self, pandas_method: dispatch_method_type):
+        """
+        A decorator to register a finalizer for a specific pandas method.
+
+        Parameters
+        ----------
+        pandas_method : callable or str
+            A pandas method, like :meth:`pandas.concat`, that this finalizer
+            should be used for. The function being decorated will be called
+            with the relevant arguments (typically the output and the source NDFrame).
+            When `NDFrame.__finalize__` is called as a result of `pandas_method`,
+            the registered finalizer will be called.
+        """
+
+        def decorate(func):
+            # TODO: warn of collisions?
+            dispatch[key_of(pandas_method)][self.name] = func
+
+            @wraps(func)
+            def wrapper(*args, **kwargs):
+                return func(*args, **kwargs)
+
+            return wrapper
+
+        return decorate
+
+
+def default_finalizer(new: "NDFrame", other: Any, *, name: str):
+    """
+    The default finalizer when this method, attribute hasn't been overridden.
+
+    This copies the ``_metadata`` attribute from ``other`` to ``self``, modifying
+    ``self`` inplace.
+
+    Parameters
+    ----------
+    new : NDFrame
+        The newly created NDFrame being finalized.
+    other : NDFrame
+        The source NDFrame attributes will be extracted from.
+    """
+    object.__setattr__(new, name, getattr(other, name, None))
+
+
+# ----------------------------------------------------------------------------
+# Pandas Internals.
+
+
+def ndframe_finalize(new: "NDFrame", other: Any, method: dispatch_method_type):
```

Comment:
Going back to previous comment, if you use `FrameOrSeries` here it is a TypeVar, so you could type it like `def ndframe_finalize(new: FrameOrSeries, ...) -> FrameOrSeries` which would guarantee the type of new and the type of the return value are the same (i.e. function becomes generic)